### PR TITLE
ci: Add virtio-fs support

### PR DESCRIPTION
**Pull request by chavafg**
Add configuration option to use virtio-fs.
We will currently use nemu for testing the
virtio-fs support.

Depends-on: github.com/kata-containers/runtime#1639

Fixes: #1536.

Signed-off-by: Salvador Fuentes <salvador.fuentes@intel.com>

### CODE REVIEW

**Comment by ganeshmaharaj** [comment]
Should we add the command to enable on the system too? `sysctl nr.huge_pages = 1024`

**Comment by chavafg** [comment]
@ganeshmaharaj addressed your comment 
/test-nemu

**Comment by chavafg** [comment]
```
[4] Running command '/usr/bin/docker [docker run --cidfile /tmp/cid844578869/H98I3diZOOuuXmiCkd0939o8fQniT4 --runtime kata-runtime --name H98I3diZOOuuXmiCkd09
39o8fQniT4 -t busybox tty]'
[4] .
[4] command failed error 'exit status 125'
[4] [docker run --cidfile /tmp/cid844578869/H98I3diZOOuuXmiCkd0939o8fQniT4 --runtime kata-runtime --name H98I3diZOOuuXmiCkd0939o8fQniT4 -t busybox tty]
[4] Timeout: 120 seconds
[4] Exit Code: 125
[4] Stdout:
[4] Stderr: docker: Error response from daemon: OCI runtime create failed: qemu-system-x86_64_virt: -object memory-backend-file,id=dimm1,size=2048M,mem-path=/
dev/hugepages,share=on,prealloc=on: unable to map backing store for guest RAM: Cannot allocate memory: unknown.
[4]
```

Seems like we have memory issues when running with virtiofs, maybe because of hugepages?
I can reproduce locally. 
If I modify the `default_memory` from `2048` to `512`, I can run more tests, but at some point I get the same error again. 

/cc @egernst @sboeuf @ganeshmaharaj

**Comment by sboeuf** [comment]
@ganeshmaharaj is already aware that hugepages are actually not needed, which is why he has a patch in flight I think. If the issue comes from hugepages, this will be solved by not using them because they're not needed per se.

**Comment by jcvenegas** [comment]
@chavafg hi, any update?

**Comment by jodh-intel** [comment]
Restarted Nemu test which failed with:

```
21:48:06 [4] • Failure in Spec Teardown (AfterEach) [125.054 seconds]
21:48:06 [4] load
21:48:06 [4] /tmp/jenkins/workspace/kata-containers-tests-ubuntu-nemu/go/src/github.com/kata-containers/tests/integration/docker/load_test.go:15
21:48:06 [4]   load with docker [AfterEach]
21:48:06 [4]   /tmp/jenkins/workspace/kata-containers-tests-ubuntu-nemu/go/src/github.com/kata-containers/tests/integration/docker/load_test.go:36
21:48:06 [4]     load a container
21:48:06 [4]     /tmp/jenkins/workspace/kata-containers-tests-ubuntu-nemu/go/src/github.com/kata-containers/tests/integration/docker/load_test.go:37
21:48:06 [4]       should load image
21:48:06 [4]       /tmp/jenkins/workspace/kata-containers-tests-ubuntu-nemu/go/src/github.com/kata-containers/tests/integration/docker/load_test.go:38
21:48:06 [4] 
21:48:06 [4]       Expected
21:48:06 [4]           <int>: -1
21:48:06 [4]       to equal
21:48:06 [4]           <int>: 0
21:48:06 [4] 
21:48:06 [4]       /tmp/jenkins/workspace/kata-containers-tests-ubuntu-nemu/go/src/github.com/kata-containers/tests/integration/docker/load_test.go:31
```

Restarted the initrd CI failed with this rather weird set of messages:

```
22:21:37 not ok 25 ctr execsync std{out,err}
22:21:37 # (in test file ctr.bats, line 946)
22:21:37 #   `[ "$status" -eq 0 ]' failed
22:21:37 # time="2019-05-15T21:21:06Z" level=error msg="File descriptor 3 (pipe:[773499]) leaked on pvdisplay invocation. Parent PID 57196: /tmp/jenkins/workspace/kata-con
22:21:37 # File descriptor 6 (/dev/mapper/control) leaked on pvdisplay invocation. Parent PID 57196: /tmp/jenkins/workspace/kata-con
22:21:37 #   Failed to find physical volume "/dev/sdb".
22:21:37 # " error="exit status 5" 
22:21:37 # 0
22:21:37 # time="2019-05-15 21:21:23.488458146Z" level=debug msg="found valid runtime 'runc' for runtime_path '/usr/local/bin/kata-runtime'
22:21:37 # " 
```

**Comment by chavafg** [comment]
seems like docker memory tests with nemu (which use virtiofs in this PR) are getting hanged. need to check furtherly

File: .ci/install_runtime.sh [hunk 1]
```diff
@@ -118,3 +118,17 @@ if [ "$KATA_EXPERIMENTAL_FEATURES" = true ]; then
 	feature="newstore"
 	sudo sed -i -e "s|^experimental.*$|experimental=[ \"$feature\" ]|" "${runtime_config_path}"
 fi
+
+# Enable virtiofs if VIRTIO_FS is set to true
+# currently we use nemu for virtiofs testing
+if [ "$VIRTIO_FS" = true ] && [ "$KATA_HYPERVISOR" = "nemu" ]; then
+	echo "Configure virtio-fs on kata-runtime config file"
+	sudo crudini --set "$runtime_config_path" hypervisor.qemu virtio_fs_daemon "\"/usr/local/bin/virtiofsd-${arch}\""
```
Comment (ganeshmaharaj) [inline .ci/install_runtime.sh]: After https://github.com/kata-containers/tests/pull/1595, maybe we can remove the hardcoding of the virtiofsd binary and use `${KATA_NEMU_DESTDIR}`

File: .ci/install_runtime.sh [hunk 2]
```diff
@@ -118,3 +118,17 @@ if [ "$KATA_EXPERIMENTAL_FEATURES" = true ]; then
...
+	sudo crudini --set "$runtime_config_path" hypervisor.qemu shared_fs "\"virtio-fs\""
+	sudo crudini --set "$runtime_config_path" hypervisor.qemu enable_hugepages "true"
```
Comment (ganeshmaharaj) [inline .ci/install_runtime.sh]: We should not longer need this. File based backend patch landed today and that should automatically take care of this part.

**Comment by chavafg** [comment]
@ganeshmaharaj now the tests are getting hanged on the cri-o tests when trying to create a container inside a pod. I just see timeout error on the kata logs, but the lasts messages are from `systemd` on `kata-proxy` log.

I tried to reproduce locally, these are the logs:

```
May 31 18:15:22 virtiofs-tests kata-proxy[9837]: time="2019-05-31T18:15:22.924103968Z" level=info msg="time=\"2019-05-31T18:15:22.907783871Z\" level=debug m
sg=\"updating cpuset cgroup\" debug_console=false name=kata-agent path=/sys/fs/cgroup/cpuset/Burstable/pod_123-456/crio-77def99ec899addc7a943b30dc3643dcadee
1ea416acb4c3836e91d6833315bb/cpuset.cpus pid=78 sandbox=77def99ec899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb source=agent\n" name=kata-proxy pid
=9837 sandbox=77def99ec899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb source=agent
May 31 18:15:22 virtiofs-tests kata-proxy[9837]: time="2019-05-31T18:15:22.939007715Z" level=info msg="[   41.331225] systemd[1]: libmount event [rescan: ye
s]\n" name=kata-proxy pid=9837 sandbox=77def99ec899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb source=agent
May 31 18:15:22 virtiofs-tests kata-proxy[9837]: time="2019-05-31T18:15:22.941015935Z" level=info msg="[   41.333207] systemd[1]: run.mount: Failed to load
configuration: No such file or directory\n" name=kata-proxy pid=9837 sandbox=77def99ec899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb source=agent
May 31 18:15:22 virtiofs-tests kata-proxy[9837]: time="2019-05-31T18:15:22.941307438Z" level=info msg="[   41.333560] systemd[1]: run-libcontainer.mount: Fa
iled to load configuration: No such file or directory\n" name=kata-proxy pid=9837 sandbox=77def99ec899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb s
ource=agent
May 31 18:15:22 virtiofs-tests kata-proxy[9837]: time="2019-05-31T18:15:22.941628441Z" level=info msg="[   41.333843] systemd[1]: run-libcontainer-77def99ec
899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb.mount: Failed to load configuration: No such file or directory\n" name=kata-proxy pid=9837 sandbox=7
7def99ec899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb source=agent
May 31 18:15:22 virtiofs-tests kata-proxy[9837]: time="2019-05-31T18:15:22.942679752Z" level=info msg="[   41.334818] systemd[1]: run-libcontainer-77def99ec
899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb-a02e080a289b62030927d519256cac214b99e45839eec6aa0064c0c47691d298.mount: Failed to load configuration
: No such file or directory\n" name=kata-proxy pid=9837 sandbox=77def99ec899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb source=agent
May 31 18:15:22 virtiofs-tests kata-proxy[9837]: time="2019-05-31T18:15:22.943669561Z" level=info msg="[   41.335223] systemd[1]: run-libcontainer-77def99ec
899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb-a02e080a289b62030927d519256cac214b99e45839eec6aa0064c0c47691d298-runc.U8dhRV.mount: Changed dead ->
mounted\n" name=kata-proxy pid=9837 sandbox=77def99ec899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb source=agent
May 31 18:15:22 virtiofs-tests kata-proxy[9837]: time="2019-05-31T18:15:22.944318168Z" level=info msg="[   41.336597] systemd[1]: run.mount: Collecting.\n"
name=kata-proxy pid=9837 sandbox=77def99ec899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb source=agent
May 31 18:15:22 virtiofs-tests kata-proxy[9837]: time="2019-05-31T18:15:22.944473069Z" level=info msg="[   41.336836] systemd[1]: run-libcontainer.mount: Co
llecting.\n" name=kata-proxy pid=9837 sandbox=77def99ec899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb source=agent
May 31 18:15:22 virtiofs-tests kata-proxy[9837]: time="2019-05-31T18:15:22.944748472Z" level=info msg="[   41.336969] systemd[1]: run-libcontainer-77def99ec
899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb.mount: Collecting.\n" name=kata-proxy pid=9837 sandbox=77def99ec899addc7a943b30dc3643dcadee1ea416acb
4c3836e91d6833315bb source=agent
May 31 18:15:22 virtiofs-tests kata-proxy[9837]: time="2019-05-31T18:15:22.945411479Z" level=info msg="[   41.337324] systemd[1]: run-libcontainer-77def99ec
899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb-a02e080a289b62030927d519256cac214b99e45839eec6aa0064c0c47691d298.mount: Collecting.\n" name=kata-pro
xy pid=9837 sandbox=77def99ec899addc7a943b30dc3643dcadee1ea416acb4c3836e91d6833315bb source=agent
May 31 18:19:22 virtiofs-tests crio[9590]: time="2019-05-31 18:19:22.718811071Z" level=error msg="Container creation timeout (4m0s)"

```

any idea?

**Comment by egernst** [comment]
@chavafg can we just use the nemu hypervisor with its default toml now?  Is this PR still needed?

**Comment by chavafg** [comment]
I need to change the PR to now use the `configuration-nemu.toml` instead of the default one.

**Comment by ganeshmaharaj** [comment]
> @ganeshmaharaj now the tests are getting hanged on the cri-o tests when trying to create a container inside a pod. I just see timeout error on the kata logs, but the lasts messages are from `systemd` on `kata-proxy` log.
> 
> I tried to reproduce locally, these are the logs:
> 
> ```
> May 31 18:15:22 virtiofs-tests kata-proxy[9837]: time="2019-05-31T18:15:22.924103968Z" level=info msg="time=\"2019-05-31T18:15:22.907783871Z\" level=debug m
> sg=\"updating cpuset cgroup\" debug_console=false name=kata-agent path=/sys/fs/cgroup/cpuset/Burstable/pod_123-456/crio-77def99ec899addc7a943b30dc3643dcadee
<snip>
> May 31 18:19:22 virtiofs-tests crio[9590]: time="2019-05-31 18:19:22.718811071Z" level=error msg="Container creation timeout (4m0s)"
> ```
> 
> any idea?

I seem to  have missed it. is this still an issue? I just tested and it seems to be working fine. and we have fixed the hotplug issue. Once the patch lands in runtime, we can turn that test back on as well.

**Comment by jodh-intel** [comment]
Greeeeeeeeeen!

**Pull request merged**